textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/join/SimilarityJoinPredicate.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    public String getIDAttributeName() {
+        return SchemaConstants._ID;
+    }
+
+    @Override
+    public String getJoinAttributeName() {
+        return joinAttributeName;
+    }
+    
+    @Override
+    public Schema generateOutputSchema(Schema innerOperatorSchema, Schema outerOperatorSchema) throws DataFlowException {
+        List<Attribute> outputAttributeList = new ArrayList<>();
+        
+        // add _ID field first
+        outputAttributeList.add(SchemaConstants._ID_ATTRIBUTE);
+        
+        for (Attribute attr : innerOperatorSchema.getAttributes()) {
+            String attrName = attr.getFieldName();
+            FieldType attrType = attr.getFieldType();
+            // ignore _id, spanList, and payload
+            if (attrName.equals(SchemaConstants._ID) || attrName.equals(SchemaConstants.SPAN_LIST) 
+                    || attrName.equals(SchemaConstants.PAYLOAD)) {
+                continue;
+            }
+            outputAttributeList.add(new Attribute(INNER_PREFIX + attrName, attrType));
+        }
+        for (Attribute attr : outerOperatorSchema.getAttributes()) {
+            String attrName = attr.getFieldName();
+            FieldType attrType = attr.getFieldType();
+            // ignore _id, spanList, and payload
+            if (attrName.equals(SchemaConstants._ID) || attrName.equals(SchemaConstants.SPAN_LIST) 
+                    || attrName.equals(SchemaConstants.PAYLOAD)) {
+                continue;
+            }
+            outputAttributeList.add(new Attribute(OUTER_PREFIX + attrName, attrType));
+        }
+        
+        // add spanList field
+        outputAttributeList.add(SchemaConstants.SPAN_LIST_ATTRIBUTE);
+
+        // add payload field if one of them contains payload
+        if (innerOperatorSchema.containsField(SchemaConstants.PAYLOAD) || 
+                outerOperatorSchema.containsField(SchemaConstants.PAYLOAD)) {
+            outputAttributeList.add(SchemaConstants.PAYLOAD_ATTRIBUTE);
+        }
+        
+        return new Schema(outputAttributeList.stream().toArray(Attribute[]::new));
+    }
+
+    @Override
+    public ITuple joinTuples(ITuple outerTuple, ITuple innerTuple, Schema outputSchema) throws DataFlowException {        
+        if (similarity == 0) {
+            return null;
+        }
+        
+        // get the span list only with the joinAttributeName
+        List<Span> innerRelevantSpanList = ((ListField<Span>) innerTuple.getField(SchemaConstants.SPAN_LIST))
+                .getValue().stream().filter(span -> span.getFieldName().equals(joinAttributeName)).collect(Collectors.toList());
+        List<Span> outerRelevantSpanList = ((ListField<Span>) outerTuple.getField(SchemaConstants.SPAN_LIST))
+                .getValue().stream().filter(span -> span.getFieldName().equals(joinAttributeName)).collect(Collectors.toList());
+        
+        // get a set of span's values (since multiple spans may have the same value)
```

Comment:
Make the comment more clear.  Do you mean we want to remove duplicates?  Also "span's values" -> "span values".